src/rewrite_reduce.cpp

```diff
@@ -36,43 +38,79 @@ inline namespace MIGRAPHX_INLINE_NS {
+        auto axis = op["axis"].to<std::int64_t>();
 
-        auto max  = m.insert_instruction(ins, make_op("reduce_max", {{"axes", {axis}}}), input);
-        auto maxb = m.insert_instruction(
+        auto input = ins->inputs().front();
+        auto max   = m.insert_instruction(ins, make_op("reduce_max", {{"axes", {axis}}}), input);
+        auto maxb  = m.insert_instruction(
             ins, make_op("multibroadcast", {{"out_lens", input->get_shape().lens()}}), max);
         auto sub  = m.insert_instruction(ins, make_op("sub"), input, maxb);
         auto exp  = m.insert_instruction(ins, make_op("exp"), sub);
         auto sum  = m.insert_instruction(ins, make_op("reduce_sum", {{"axes", {axis}}}), exp);
         auto sumb = m.insert_instruction(
             ins, make_op("multibroadcast", {{"out_lens", input->get_shape().lens()}}), sum);
-        auto div = m.insert_instruction(ins, make_op("div"), exp, sumb);
+        m.replace_instruction(ins, make_op("div"), exp, sumb);
+    }
+};
 
-        if(full_precision and requires_upcast)
-        {
-            m.replace_instruction(ins, make_op("convert", {{"target_type", input_type}}), div);
-        }
-        else
+struct find_softmax_base_ops
+{
+    bool full_precision;
+
+    auto matcher() const { return match::softmax(); }
+
+    void apply(module& m, const match::matcher_result& r) const
+    {
+        auto div             = r.result;
+        auto inp             = r.instructions["x"];
+        auto inp_type        = inp->get_shape().type();
+        auto requires_upcast = not contains({shape::float_type, shape::double_type}, inp_type);
+
+        if(not requires_upcast)
+            return;
+
+        auto softmax_inss = find_instructions_between(inp, div, &m);
+
+        // add converts back to original type for all intermediate and final outputs
+        for(auto i : softmax_inss)
         {
-            m.replace_instruction(ins, div);
+            auto outs = i->outputs();
+
+            std::unordered_set<instruction_ref> downcast_outputs;
+            std::copy_if(outs.begin(),
+                         outs.end(),
+                         std::inserter(downcast_outputs, downcast_outputs.begin()),
+                         [&](auto o) { return not contains(softmax_inss, o); });
+
+            if(downcast_outputs.empty())
+                continue;
+
+            auto i_down = m.insert_instruction(
+                std::next(i), make_op("convert", {{"target_type", inp_type}}), i);
+
+            for(auto o : downcast_outputs)
+            {
+                instruction::replace_argument(o, i, i_down);
+            }
         }
+
+        std::vector<instruction_ref> sorted_softmax_inss(softmax_inss.begin(), softmax_inss.end());
+        std::sort(sorted_softmax_inss.begin(),
+                  sorted_softmax_inss.end(),
+                  [&](instruction_ref x, instruction_ref y) {
+                      return std::distance(m.begin(), x) < std::distance(m.begin(), y);
```

Comment:
This could be very slow for large models, and its called multiple times during the sort.